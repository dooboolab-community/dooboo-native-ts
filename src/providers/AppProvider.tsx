import React, { useReducer } from 'react';
import { AppContext } from '../contexts';
import { IUser } from '../types';
import { ThemeType } from '../theme';

const AppConsumer = AppContext.Consumer;

interface IAction {
  type: 'reset-user' | 'set-user' | 'change-theme-mode';
  payload: any;
}

interface IProps {
  navigation?: any;
  children?: any;
}

export interface IState {
  user: IUser;
  theme: ThemeType;
}

const initialState: IState = {
  theme: ThemeType.LIGHT,
  user: {
    displayName: '',
    age: 0,
    job: '',
  },
};

const reducer = (state: IState, action: IAction) => {
  switch (action.type) {
    case 'change-theme-mode':
      return { ...state, theme: action.payload.theme };
    case 'reset-user':
      return { ...state, user: initialState.user };
    case 'set-user':
      return { ...state, user: action.payload };
  }
};

function AppProvider(props: IProps) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = { state, dispatch };

  return (
    <AppContext.Provider value={value}>
      {props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
