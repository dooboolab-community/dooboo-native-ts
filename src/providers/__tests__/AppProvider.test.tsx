import * as React from 'react';

import { AppProvider, useAppContext } from '../AppProvider';
import { Button, Text, View } from 'react-native';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const FakeChild = (): React.ReactElement => {
  const { state, resetUser } = useAppContext();

  return (
    <View>
      <Text testID="TEXT">{JSON.stringify(state, null, 2)}</Text>
      <Button
        testID="BUTTON"
        onPress={(): void => {
          resetUser();
        }}
        title="Button"
      />
    </View>
  );
};

describe('[AppProvider] rendering test', () => {
  let json: renderer.ReactTestRendererJSON;
  const component = (
    <AppProvider>
      <FakeChild />
    </AppProvider>
  );

  it('component and snapshot matches', () => {
    json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toBeTruthy();
  });
});

// TODO: add more interaction test, refer to ThemeProvider test
