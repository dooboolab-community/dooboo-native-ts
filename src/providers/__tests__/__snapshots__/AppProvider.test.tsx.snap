// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AppProvider] rendering test should match component and snapshot 1`] = `
<View>
  <Text
    testID="TEXT"
  >
    {
  "user": null
}
  </Text>
  <Button
    onPress={[Function]}
    testID="BUTTON"
    title="Button"
  />
  <Button
    onPress={[Function]}
    testID="BUTTON_NOT_VALID"
    title="Button"
  />
</View>
`;
